File: python/alibabacloud_tea_openapi/websocket_utils/general.py
import json
from typing import Any, Dict

from darabonba.websocket import (
    AbstractWebSocketHandler,
    WebSocketMessage,
    WebSocketMessageType,
    WebSocketSessionInfo,
)

from ._constants import ERR_USE_RAW_MESSAGE


GeneralMessage = Dict[str, Any]


class GeneralWebSocketHandler(AbstractWebSocketHandler):
    def handle_general_message(self, session: WebSocketSessionInfo, message: GeneralMessage) -> None:
        raise ERR_USE_RAW_MESSAGE


def new_general_message(body: str) -> GeneralMessage:
    return {'body': body}


def parse_general_message(message: WebSocketMessage) -> GeneralMessage:
    if message.type == WebSocketMessageType.Binary:
        return {
            'body': message.payload,
            'format': 'binary',
        }
    try:
        return {
            'body': json.loads(message.payload.decode('utf-8')),
            'format': 'text',
        }
    except (json.JSONDecodeError, UnicodeDecodeError):
        return {
            'body': message.payload,
            'format': 'text',
        }


def general_message_to_json(message: GeneralMessage) -> bytes:
    return json.dumps(message, separators=(',', ':')).encode('utf-8')

File: ts/src/websocketUtils/general.ts
import * as $dara from '@darabonba/typescript';
import { ErrUseRawMessage } from './awap';

export type GeneralMessageFormat = 'text' | 'binary';

export interface GeneralMessage {
  body?: any;
  format?: GeneralMessageFormat;
}

export interface GeneralWebSocketHandler extends $dara.WebSocketHandler {
  handleGeneralMessage(session: $dara.WebSocketSessionInfo, message: GeneralMessage): Promise<void> | void;
}

export class AbstractGeneralWebSocketHandler extends $dara.AbstractWebSocketHandler implements GeneralWebSocketHandler {
  handleGeneralMessage(_session: $dara.WebSocketSessionInfo, _message: GeneralMessage): void {
    throw ErrUseRawMessage;
  }
}

export function newGeneralMessage(body: string): GeneralMessage {
  return { body };
}

export function parseGeneralMessage(message: $dara.WebSocketMessage): GeneralMessage {
  if (message.type === $dara.WebSocketMessageType.Binary) {
    return {
      body: message.payload,
      format: 'binary',
    };
  }

  try {
    return {
      body: JSON.parse(message.payload.toString('utf8')),
      format: 'text',
    };
  } catch (_err) {
    return {
      body: message.payload,
      format: 'text',
    };
  }
}

export function generalMessageToJSON(message: GeneralMessage): Buffer {
  return Buffer.from(JSON.stringify(message));
}

export { ErrUseRawMessage };
